lib/srv/db/access_test.go

```diff
@@ -778,20 +778,29 @@ func TestAccessMongoDB(t *testing.T) {
 	}{
 		{
 			name: "client without compression",
-			opts: options.Client(),
+			opts: options.Client().
+				// Add extra time so the test won't time out when running in parallel.
+				SetServerSelectionTimeout(10 * time.Second),
 		},
 		{
 			name: "client with compression",
-			opts: options.Client().SetCompressors([]string{"zlib"}),
+			opts: options.Client().
+				SetServerSelectionTimeout(10 * time.Second).
+				SetCompressors([]string{"zlib"}),
 		},
 	}
 
 	// Execute each scenario on both modern and legacy Mongo servers
 	// to make sure legacy messages are also subject to RBAC.
 	for _, test := range tests {
+		test := test
 		for _, serverOpt := range serverOpts {
+			serverOpt := serverOpt
 			for _, clientOpt := range clientOpts {
+				clientOpt := clientOpt
 				t.Run(fmt.Sprintf("%v/%v/%v", serverOpt.name, clientOpt.name, test.desc), func(t *testing.T) {
+					t.Parallel()
```

Comment:
I wonder if we'd get better performance if we grouped this up and maybe parallelized this at the first layer in the loop instead?

Reply:
@espadolini Quick test on my Mac.
non-parallel version:
```
% time go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db
ok      github.com/gravitational/teleport/lib/srv/db    542.742s
go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db  808.37s user 60.45s system 157% cpu 9:10.02 total
```

parallel in the most inner loop
```
% time go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db
ok      github.com/gravitational/teleport/lib/srv/db    99.992s
go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db  557.41s user 84.62s system 594% cpu 1:48.07 total
```

I'm not sure what did you mean by "group this up" so here are my best guesses.

Move `t.Run()` to the most outer loop, so two most inner loops run inside inside it.

```go
	for _, test := range tests {
		test := test
		t.Run(fmt.Sprintf("%v", test.desc), func(t *testing.T) {
			t.Parallel()
			for _, serverOpt := range serverOpts {
				serverOpt := serverOpt
				for _, clientOpt := range clientOpts {
					clientOpt := clientOpt
```

The problem here is that now each `t.Run()` runs 4 test cases, so in case of failure we won't know which of those 4 failed. 

```
 % time go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db
ok      github.com/gravitational/teleport/lib/srv/db    53.812s
go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db  287.35s user 32.47s system 518% cpu 1:01.64 total
```

The same as above, but with each test case in separate `t.Run()`. This fixes the problem described above.

```go
	for _, test := range tests {
		test := test
		t.Run(fmt.Sprintf("%v", test.desc), func(t *testing.T) {
			t.Parallel()
			for _, serverOpt := range serverOpts {
				serverOpt := serverOpt
				for _, clientOpt := range clientOpts {
					clientOpt := clientOpt

					t.Run(fmt.Sprintf("%v/%v", serverOpt.name, clientOpt.name), func(t *testing.T) {
```

```
% time go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db
ok      github.com/gravitational/teleport/lib/srv/db    149.265s
go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db  822.25s user 78.82s system 578% cpu 2:35.76 total
```

Two `t.Run()` with `t.Parallel()`.

```go
	for _, test := range tests {
		test := test
		t.Run(fmt.Sprintf("%v", test.desc), func(t *testing.T) {
			t.Parallel()
			for _, serverOpt := range serverOpts {
				serverOpt := serverOpt
				for _, clientOpt := range clientOpts {
					clientOpt := clientOpt

					t.Run(fmt.Sprintf("%v/%v", serverOpt.name, clientOpt.name), func(t *testing.T) {
						t.Parallel()
```

```
% time go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db
ok      github.com/gravitational/teleport/lib/srv/db    90.712s
go test -race -run TestAccessMongoDB -count=3 ./lib/srv/db  536.56s user 79.32s system 636% cpu 1:36.74 total
```

Reply:
I don't understand why the version with the parallelized outer loop and serialized inner loop with subtests ends up being so much slower than the other ones :\

The time save on parallelizing just the outer loop is pretty insane tho, 100 to 50 seconds - maybe it's worth doing it regardless?

Reply:
I know why. My mistake while testing: https://github.com/gravitational/teleport/blob/3e72b2e530a4c6f7ab8eb51fd2ec458b00e5a40f/lib/srv/db/access_test.go#L817-L830
These `return` statements were terminating the test after a few iterations.
Updated time (with `continue` instead of `return`):
```
FAIL
FAIL    github.com/gravitational/teleport/lib/srv/db    174.194s
FAIL
go test -race -run TestAccessMongoDB -timeout=30m -count=3 ./lib/srv/db  596.03s user 206.98s system 445% cpu 3:00.15 total
```

I noticed that the test server can be reused between tests, as it state is not modified. I moved it one loop up. Here are the results:
```
% time go test -race -run TestAccessMongoDB -timeout=30m -count=3 ./lib/srv/db
ok      github.com/gravitational/teleport/lib/srv/db    81.794s
go test -race -run TestAccessMongoDB -timeout=30m -count=3 ./lib/srv/db  488.24s user 66.11s system 634% cpu 1:27.30 total
```

Looks like that saved a few seconds. I updated the PR.

Reply:
540 to 80s then? 🥳